frontend/catalyst/api_extensions/callbacks.py

```diff
@@ -39,6 +41,63 @@
 
 
 ## API ##
+def accelerate(func=None, dev=None):
+    """Seamless integration with jax.jit
+
+    The accelerate function can be used to automatically apply `jax.jit` to
+    the input function `func` and be used to specify the device with the `dev`
+    keyword argument. If `dev` is None, then the default device as determined
+    by JAX is used.
+
+    Can be used in the following modes:
+
+
+    ```py
+    @accelerate
+    def identity(x):
+      return x
+
+    @qml.qjit
+    def qjitted_fn(x):
+      return x
+    ```
+
+    ```
+    @accelerate(dev=jax.devices()[0])
+    def identity(x):
+      return x
+
+    @qml.qjit
+    def qjitted_fn(x):
+      return x
+    ```
+
+    or
+
+    ```
+    @qml.qjit
+    def qjitted_fn(x):
+      return accelerate(lambda x: x)
+    ```
+    """
+
+    if isinstance(func, Callable):
+
+        @functools.wraps(func)
+        def defer(*args, **kwargs):
+            # Make abstract variables from input tracers.
+            absargs, abskwargs = tree_map(shaped_abstractify, (args, kwargs))
+            # Find the shape of the return value
+            _, returnshape = jax.make_jaxpr(func, return_shape=True)(*absargs, **abskwargs)
+            jitted_fn = jax.jit(func)
+            return jax_jit_callback(jitted_fn, returnshape, device=dev)(*args, **kwargs)
+
+        return defer
+
+    kwargs = copy.copy(locals())
+    kwargs.pop("func")
```

Comment:
It's safer to get the locals right at the top, because code above can generate new locals. Inverting the condition means the partial application can also be handled up top.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/805/commits/ece8aa8d425ea2f691f9aef35143f1bab821faa3